### DISCUSSION TITLE
MongoDB 6.0 Support / Encryption / Change Streams / and more

### DISCUSSION THREAD

**Discussion by wreiske** [meteor/meteor#12092] [Feature Requests]
Starting a discussion about MongoDB 6.0. 

https://www.mongodb.com/blog/post/big-reasons-upgrade-mongodb-6-0
https://www.mongodb.com/blog/post/change-streams-mongodb-6-0-support-pre-post-image-retrieval-ddl-operations
https://www.mongodb.com/blog/post/mongodb-releases-queryable-encryption-preview
https://www.mongodb.com/docs/v6.0/reference/read-concern-snapshot/

Key points:

1. Even more support for working with time series data
2. A better way to build event-driven architectures (better change streams & DDL!)
3. Deeper insights from enriched queries
4. More operators, less work
5. More resilient operations
6. Additional data security and operational efficiency
7. A smoother search experience and seamless data sync

I have a test app running with MongoDB 6.0 and it's working fine so far in development! Haven't had any errors yet 🤞

**Comment by radekmie** [comment]
You mean _bundling_ MongoDB 6.0 or _making it work_ with it? The driver is most likely fine (version [here](https://github.com/meteor/meteor/blob/10a1e014171850d1d537449866a3b9216064364c/packages/npm-mongo/package.js#L11)), even though [MongoDB says otherwise](https://www.mongodb.com/docs/drivers/node/v4.3/compatibility/). They often mark it as incompatible only because it can't handle the new features. _(As you said - it works.)_

Having said that, I just read the release notes of the driver from 4.4 to 4.8 and... [4.7 looks _extremely_ interesting](https://github.com/mongodb/node-mongodb-native/releases/tag/v4.7.0):

> **MongoClient.connect is now optional**
> If an operation is run before MongoClient.connect is called by the client, the driver will now automatically connect along with that first operation. This makes the repl experience much more streamlined, going right from client construction to your first insert or find. However, MongoClient.connect can still be called manually and remains useful for learning about misconfiguration (auth, server not started, connection string correctness) early in your application's startup.

I'll look into that, as we may have one global promise less to migrate (cc @zodern, @fredmaiaarantes, @henriquealbert).

**Reply by henriquealbert** [reply]
That would be nice to have. Thanks for cc'ing me

**Reply by radekmie** [reply]
I've tried it out, and the results are somehow interesting.
1. [This code](https://github.com/meteor/meteor/blob/release-2.7.4/packages/mongo/mongo_driver.js#L201-L206) is not needed, as the [`hello` command](https://www.mongodb.com/docs/manual/reference/command/hello/) is supported in all versions starting from 3.6 (3.6.21 to be exact, but that's still an _old_ version), and the [current driver version](https://github.com/meteor/meteor/blob/release-2.7.4/packages/npm-mongo/package.js) (4.3.1) doesn't support anything below that anyway ([docs](https://www.mongodb.com/docs/drivers/node/v4.3/compatibility/)).
2. Getting rid of all of the `await`s makes [all of the constructors mentioned in the PR](https://github.com/meteor/meteor/pull/12028) synchronous again. However, there's a catch: the connection error is not synchronous anymore.

I'm checking what we could do about it, but if we're fine with that, then it simplifies _a lot_.

**Reply by radekmie** [reply]
I published my changes in https://github.com/meteor/meteor/pull/12097.

**Reply by henriquealbert** [reply]
with this update, would we roll back 2.7.4 changes?

**Reply by radekmie** [reply]
@henriquealbert The new APIs will be there, but none of the constructors will be asynchronous anymore.

**Comment by vlasky** [comment]
I am most excited by the enhancement to change streams that now allows users to retrieve the before and after state of the modified document. With these enhancements, it seems we can eliminate the need for Meteor's MongoDB oplog tailing code.

MongoDB's oplog format is undocumented and Meteor developers have only been able to implement support through reverse-engineering - a painstaking process.

**Reply by radekmie** [reply]
> With these enhancements, it seems we can eliminate the need for Meteor's MongoDB oplog tailing code.

I think it's not that trivial. Enabling pre-images incurs a few significant limitations ([docs](https://www.mongodb.com/docs/manual/reference/method/db.collection.watch/#std-label-db.collection.watch-change-streams-pre-and-post-images-example)), most importantly cutting the document size limit in half and incurring an additional performance overhead.

Also, I haven't checked it in 5.1 and later versions, but there was a significant cost of creating e.g., 1000 of change streams. They were heavily optimized since then, but whether or not it'll handle 10k or 100k change streams is something we'd need to measure before going anywhere with it.
